providers/defltprov.c

```diff
@@ -331,134 +331,138 @@ static OSSL_ALGORITHM exported_ciphers[OSSL_NELEM(deflt_ciphers)];
+    { "PBKDF2", "provider=default", kdf_pbkdf2_functions },
+    { "SSHKDF", "provider=default", kdf_sshkdf_functions },
+    { "X963KDF", "provider=default", kdf_x963_kdf_functions },
+    { "TLS1-PRF", "provider=default", kdf_tls1_prf_functions },
+    { "KBKDF", "provider=default", kdf_kbkdf_functions },
 #ifndef OPENSSL_NO_CMS
-    { "X942KDF", "default=yes", kdf_x942_kdf_functions },
+    { "X942KDF", "provider=default", kdf_x942_kdf_functions },
 #endif
 #ifndef OPENSSL_NO_SCRYPT
-    { "SCRYPT:id-scrypt", "default=yes", kdf_scrypt_functions },
+    { "SCRYPT:id-scrypt", "provider=default", kdf_scrypt_functions },
 #endif
-    { "KRB5KDF", "default=yes", kdf_krb5kdf_functions },
+    { "KRB5KDF", "provider=default", kdf_krb5kdf_functions },
     { NULL, NULL, NULL }
 };
 
 static const OSSL_ALGORITHM deflt_keyexch[] = {
 #ifndef OPENSSL_NO_DH
-    { "DH:dhKeyAgreement", "default=yes", dh_keyexch_functions },
+    { "DH:dhKeyAgreement", "provider=default", dh_keyexch_functions },
 #endif
 #ifndef OPENSSL_NO_EC
-    { "X25519", "default=yes", x25519_keyexch_functions },
-    { "X448", "default=yes", x448_keyexch_functions },
+    { "X25519", "provider=default", x25519_keyexch_functions },
+    { "X448", "provider=default", x448_keyexch_functions },
 #endif
     { NULL, NULL, NULL }
 };
 
 static const OSSL_ALGORITHM deflt_signature[] = {
 #ifndef OPENSSL_NO_DSA
-    { "DSA:dsaEncryption", "default=yes", dsa_signature_functions },
+    { "DSA:dsaEncryption", "provider=default", dsa_signature_functions },
 #endif
     { NULL, NULL, NULL }
 };
 
 static const OSSL_ALGORITHM deflt_asym_cipher[] = {
-    { "RSA:rsaEncryption", "default=yes", rsa_asym_cipher_functions },
+    { "RSA:rsaEncryption", "provider=default", rsa_asym_cipher_functions },
     { NULL, NULL, NULL }
 };
 
 static const OSSL_ALGORITHM deflt_keymgmt[] = {
 #ifndef OPENSSL_NO_DH
-    { "DH:dhKeyAgreement", "default=yes", dh_keymgmt_functions },
+    { "DH:dhKeyAgreement", "provider=default", dh_keymgmt_functions },
 #endif
 #ifndef OPENSSL_NO_DSA
-    { "DSA:dsaEncryption", "default=yes", dsa_keymgmt_functions },
+    { "DSA:dsaEncryption", "provider=default", dsa_keymgmt_functions },
 #endif
-    { "RSA:rsaEncryption", "default=yes", rsa_keymgmt_functions },
+    { "RSA:rsaEncryption", "provider=default", rsa_keymgmt_functions },
 #ifndef OPENSSL_NO_EC
-    { "X25519", "default=yes", x25519_keymgmt_functions },
-    { "X448", "default=yes", x448_keymgmt_functions },
+    { "X25519", "provider=default", x25519_keymgmt_functions },
+    { "X448", "provider=default", x448_keymgmt_functions },
 #endif
     { NULL, NULL, NULL }
 };
 
+/*
+ * Unlike most algorithms in the default provider, the serializers are compatible
+ * for use in FIPS mode, and therefore have the "fips=yes" property.
+ */
 static const OSSL_ALGORITHM deflt_serializer[] = {
-    { "RSA", "default=yes,format=text,type=private",
+    { "RSA", "provider=default,fips=yes,format=text,type=private",
```

Comment:
I am confused..
When we do a query with fips=n o doesnt that mean it wont find the serializer??

Reply:
Er, I concur.  Why should the **default** provider have _anything_ with the property `fips=yes`?

Reply:
Because, "fips=yes" now means that the algorithm is compatible for use in FIPS mode. The serializers in the default provider *are* compatible and hence have the property. A query of "fips=no" means only algorithms that are not compatible for use in FIPS mode, and would rightly exclude the serializers. If you instead mean only those algorithms that are not in the FIPS *provider* then the query string should be "provider!=fips".

Reply:
I cannot think of a case where _FIPS=no_ is used.  Perhaps non-US governments might but they'll not be asking for a NIST certified algorithm and thus they won't want a FIPS serialiser.

Reply:
The absence of the fips property doesn't mean "fips=no", does it? I gotta admit that I'm a bit confused on that...

Reply:
I think knowing whether or not an algorithm is FIPS does not belong in OpenSSL; it belongs in the FIPS documents -- are you promising to keep current whenever they issue an update, for example? If it *does* end up in OpenSSL, I think it should be a separate stand-alone function, table-driven and global. It should not be part of the property lookup mechanism.

Reply:
Sorry, I don't get a point. 

We have some algorithms that are allowed according to FIPS documents. 
We have various implementations of them. Some of them are FIPS-certified.

These 2 are different properties. What does "fips=yes" mean?

Reply:
"fips=yes" means it is allowed to be used in FIPS mode. This is different to it being an algorithm that is implemented in the FIPS provider (although it includes all of those). Essentially in FIPS mode you can use all the algorithms in the FIPS provider plus the serializers.

Reply:
>"fips=yes" means it is allowed to be used in FIPS mode.

NO.  This is wrong.  What is allowed to be used in FIPS mode is *FIPS-validated implementations.*  Algorithms that are acceptable are enumerated by NIST, but you can't use a non-validated implementation in FIPS mode.  This will cause immense confusion.

Reply:
> I think knowing whether or not an algorithm is FIPS does not belong in OpenSSL; it belongs in the FIPS documents -- are you promising to keep current whenever they issue an update, for example?

You are entitled to your own view.
A fips provider is a list of certified algorithms..
e.g: https://csrc.nist.gov/projects/cryptographic-module-validation-program/Certificate/2473
It doesnt need to implement every single algorithm there is at any point in time - it can just be one algorithm if that is all that is needed.
Fips modules can also contain non fips algorithms (that can be accessed in non fips mode) if you really want.

You will also find the FIPS_mode() and FIPS_mode_set() listed in the design document.

Reply:
> You will also find the FIPS_mode() and FIPS_mode_set() listed in the design document.

And a wrapper mode for engines.  Your point being?

Reply:
Regarding engines, we chose to leave them as they are.  A wrapper proved to be much harder than anticipated, and time runeth away.  That doesn't mean that we exclude a _future_ wrapper, though.

As far as I'm personally concerned, I wouldn't include `FIPS_mode()` or `FIPS_mode_set()` at all, but it was concluded very early (in the first design meeting) that there those that will want to keep using these.

So from an application point of view, ENGINEs are still going to be there in 3.0 as before, and `FIPS_mode()` / `FIPS_mode_set()` as well.

Reply:
> NO. This is wrong. What is allowed to be used in FIPS mode is FIPS-validated implementations. Algorithms that are acceptable are enumerated by NIST, but you can't use a non-validated implementation in FIPS mode. This will cause immense confusion.

You do realise that what we are talking about here are the serializers? If you disable access to the serializers then you won't be able to:

- Save a generated private key for later use
- Send a public key to anyone
- Construct certificates
- Use TLS/CMS or any other similar protocol that transmits public keys between peers

FIPS without these capabilities is useless. AFAIK the OpenSSL FOM 2.0 did not disable access to these things.

Reply:
There's a lot of side discussion in this comment thread.  My mistake picking up one of those side tracks...
The original concern seems to be resolved as far as I can tell, so...